### ISSUE TITLE
Access to apiservers fails from any container

### ISSUE THREAD

**Issue by Bregor** [kelseyhightower/kubernetes-the-hard-way#25]
From kube-dns:

    E0713 20:38:46.910179       1 reflector.go:216] pkg/dns/dns.go:154: Failed to list *api.Endpoints: Get https://10.32.0.1/api/v1/endpoints?resourceVersion=0: dial tcp 10.32.0.1:443: i/o timeout
    E0713 20:38:46.910318       1 reflector.go:216] pkg/dns/dns.go:155: Failed to list *api.Service: Get https://10.32.0.1/api/v1/services?resourceVersion=0: dial tcp 10.32.0.1:443: i/o timeout
    I0713 20:38:46.910438       1 dns.go:172] Ignoring error while waiting for service default/kubernetes: Get https://10.32.0.1/api/v1/namespaces/default/services/kubernetes: dial tcp 10.32.0.1:443: i/o timeout. Sleeping 1s before retrying.

From random pod:

    $ kubectl run curl --image=tutum/curl:latest yes
    deployment "curl" created

    $ kubectl exec -it curl-14193874-k3esx bash

    # curl https://10.32.0.1
    curl: (7) Failed to connect to 10.32.0.1 port 443: Connection timed out

    # curl https://10.240.0.20:6443
    curl: (7) Failed to connect to 10.240.0.20 port 6443: Connection timed out

Routes:

    $ gcloud compute routes list
    NAME                            NETWORK     DEST_RANGE     NEXT_HOP                  PRIORITY
    default-route-4e889f0c670475f9  default     0.0.0.0/0      default-internet-gateway  1000
    default-route-6aca960aaec01e82  kubernetes  0.0.0.0/0      default-internet-gateway  1000
    default-route-b3f6dfcccde00a9b  kubernetes  10.240.0.0/24                            1000
    default-route-fe4c4def818434df  default     10.240.0.0/16                            1000
    kubernetes-route-10-200-0-0-24  kubernetes  10.200.0.0/24  10.240.0.30               1000
    kubernetes-route-10-200-1-0-24  kubernetes  10.200.1.0/24  10.240.0.31               1000
    kubernetes-route-10-200-2-0-24  kubernetes  10.200.2.0/24  10.240.0.32               1000

Firewall rules:

    $ gcloud compute firewall-rules list
    NAME                         NETWORK     SRC_RANGES      RULES                         SRC_TAGS  TARGET_TAGS
    default-allow-icmp           default     0.0.0.0/0       icmp
    default-allow-internal       default     10.240.0.0/16   tcp:1-65535,udp:1-65535,icmp
    default-allow-ssh            default     0.0.0.0/0       tcp:22
    kubernetes-allow-api-server  kubernetes  0.0.0.0/0       tcp:6443
    kubernetes-allow-healthz     kubernetes  130.211.0.0/22  tcp:8080
    kubernetes-allow-icmp        kubernetes  0.0.0.0/0       icmp
    kubernetes-allow-internal    kubernetes  10.240.0.0/24   tcp:0-65535,udp:0-65535,icmp
    kubernetes-allow-rdp         kubernetes  0.0.0.0/0       tcp:3389
    kubernetes-allow-ssh         kubernetes  0.0.0.0/0       tcp:22

What am I missing?

**Comment by Bregor** [comment]
Communications between "real" hosts works:

    worker0:~$ curl --cacert /var/lib/kubernetes/ca.pem  https://10.32.0.1
    Unauthorized

    worker0:~$ curl --cacert /var/lib/kubernetes/ca.pem https://10.240.0.20:6443
    Unauthorized

**Comment by sgonzalezO** [comment]
I used these commands to delete secrets on the cluster:
``
kubectl delete -f ../dns/skydns-rc.yaml (Don't know if it's necesary)
kubectl get secrets --all-namespaces
kubectl delete secrets default-token-pzwtx --namespace=kube-system
kubectl delete secrets default-token-pnotr
``
A few seconds after removing the secrets, both are recreated. 

Then I created again dns replication controller and it worked as expected.

Hope this help you.

PD: I don't know what secrets are for...

**Comment by Bregor** [comment]
@sgonzalezO thanks for suggestion, but it seems to be another case.
After recreation of default secrets and RC result still the same:

    I0714 16:29:14.814649       1 dns.go:172] Ignoring error while waiting for service default/kubernetes: Get https://10.32.0.1:443/api/v1/namespaces/default/services/kubernetes: dial tcp 10.32.0.1:443: i/o timeout. Sleeping 1s before retrying.
    E0714 16:29:14.817562       1 reflector.go:216] pkg/dns/dns.go:154: Failed to list *api.Endpoints: Get https://10.32.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.32.0.1:443: i/o timeout 
    E0714 16:29:14.817934       1 reflector.go:216] pkg/dns/dns.go:155: Failed to list *api.Service: Get https://10.32.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.32.0.1:443: i/o timeout

**Comment by robbiev** [comment]
We are also seeing this exact same issue (including it working from the host), though we are not using google cloud. Did you find a solution?

**Comment by bonovoxly** [comment]
I saw ran into this, figured I'd share.  Took a lot of `tcpdump` to hunt it out.  

In cloud platforms, you will most likely run into two issues:

- routing
- ACL/security groups for the docker pods to any other system (aka, the Kubernetes API).

When you see an error like this:
```
Failed to list *api.Service: Get https://10.32.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.32.0.1:443: i/o timeout
```

There is a very good chance it is one of the above.  When `kubedns` accesses the API server using the service address (`10.32.0.0/24`), it will actually access the API server via the `10.240.x.x:6443` address.  I believe iptables is NAT'ing 10.32.x.x:443 to 10.240.x.x:6443.  I like to cheat when discovering issues like this, so I vigorously use tools like `tcpdump`. If you `tcpdump` on the kubelet server (or API server that answers), you'll see something like:
```
12:10:35.974682 IP 10.200.0.2.48726 > 10.240.0.3.6443: Flags [S], seq 1424513382, win 29200, options [mss 1460,sackOK,TS val 36831829 ecr 0,nop,wscale 7], length 0
```

What you're seeing is kubedns performing the error listed by @Bregor .  You can reproduce the error by running `curl -k https://10.32.0.1` (just to generate traffic) in an interactive Kubernetes pod (command at the end of this post). After that, I used `kubectl describe pods` to find where it was running (what kubelet), then launched `tcpdump` on that specific kubelet.

After that, what you need to determine is if it's a routing issue or ACL/security group.  You now know which API server the request is going to from that `tcpdump` output.  `tcpdump` on that host and verify traffic arrives.  If you see no such traffic, it's most likely ACL/security group issues with your cloud provider.  Note that some restrict foreign CIDR traffic (if your VPC or whatever CIDR is 10.240.0.0/16, 10.200.0.0/16 is foreign and it may be filtering it by default). 

If in the tcpdump you see the API server is attempting to answer the request, it most likely is a routing issue.  you'll want to route the 10.200.x.x/24 traffic to the appropriate kubelet.

TLDR; vigorous use of tcpdump to determine if it's a routing issue or ACL/security group issue.

Also a handy command for launching an interactive Kubernetes pod (using CentOS instead of Ubuntu because this Docker image has `curl` and `host` commands):

```
kubectl run -i -t centos7interactive2 --restart=Never --image=centos:7 /bin/bash
```

**Comment by yawboateng** [comment]
I ran into this same issue (not using google cloud). Solved by adding route to 10.200.x.x/24 on the controller/api node. 

EX:
```
kubectl get nodes \
  --output=jsonpath='{range .items[*]}{.status.addresses[?(@.type=="InternalIP")].address} {.spec.podCIDR} {"\n"}{end}'
```
output: 
`10.240.0.30 10.200.0.0/24 `

SSH onto the API server and run:

`route add -net 10.200.0.0 netmask 255.255.255.0 gw 10.240.0.30`

**Comment by kelseyhightower** [comment]
The real fix for this issues is to ensure the `--no-source-dest-check` flag is set on EC2 instances. For example:

```
aws ec2 modify-instance-attribute \
  --instance-id ${CONTROLLER_0_INSTANCE_ID} \
  --no-source-dest-check
```

I've updated this guide with support for AWS that captures these requirements as well.
https://github.com/kelseyhightower/kubernetes-the-hard-way/blob/master/docs/01-infrastructure-aws.md

Let me know if this works.

**Comment by kelseyhightower** [comment]
This should be fixed with the latest updates to the guide. Please feel free to reopen this issue if that is not the case. See my previous comment regarding updated docs for amazon.

**Comment by bamb00** [comment]
I have the same issue timed out cannot access to apiserver from a container,

[ root@redis-master:/data ]$ **curl https://10.96.0.1:443/api**
curl: (7) Failed to connect to 10.96.0.1 port 443: Connection timed out

**Comment by bamb00** [comment]
Open new issue #128 since this is a close issue.

**Comment by lliknart** [comment]
@kelseyhightower 
I faced the same issue and when I modified the kube-proxy mode from iptables to ipvs, it worked...
